compiler/src/dotty/tools/dotc/typer/ImportInfo.scala

```diff
@@ -55,62 +56,86 @@ class ImportInfo(symf: Context => Symbol, val selectors: List[untpd.Tree],
+    myForwardMapping = SimpleIdentityMap.Empty
+    myReverseMapping = SimpleIdentityMap.Empty
     def recur(sels: List[untpd.Tree]): Unit = sels match {
       case sel :: sels1 =>
         sel match {
           case Thicket(Ident(name: TermName) :: Ident(nme.WILDCARD) :: Nil) =>
             myExcluded += name
           case Thicket(Ident(from: TermName) :: Ident(to: TermName) :: Nil) =>
-            myMapped = myMapped.updated(to, from)
+            myForwardMapping = myForwardMapping.updated(from, to)
+            myReverseMapping = myReverseMapping.updated(to, from)
             myExcluded += from
-            myOriginals += from
           case Ident(nme.WILDCARD) =>
             myWildcardImport = true
           case Ident(name: TermName) =>
-            myMapped = myMapped.updated(name, name)
-            myOriginals += name
+            myForwardMapping = myForwardMapping.updated(name, name)
+            myReverseMapping = myReverseMapping.updated(name, name)
+          case TypeBoundsTree(_, tpt) =>
+            myWildcardImport = true // details are handled separately in impliedBounds
         }
         recur(sels1)
       case nil =>
     }
     recur(selectors)
   }
 
+  private[this] var myImpliedBound: Type = null
+
+  def impliedBound(implicit ctx: Context): Type = {
+    if (myImpliedBound == null)
+      myImpliedBound = selectors.lastOption match {
+        case Some(TypeBoundsTree(_, untpd.TypedSplice(tpt))) => tpt.tpe
+        case Some(TypeBoundsTree(_, tpt)) =>
+          myImpliedBound = NoType
+          ctx.typer.typedAheadType(tpt).tpe
+        case _ => NoType
+      }
+    myImpliedBound
+  }
+
   private def implicitFlag(implicit ctx: Context) =
     if (importImplied || ctx.mode.is(Mode.FindHiddenImplicits)) ImplicitOrImpliedOrGiven
     else Implicit
 
   /** The implicit references imported by this import clause */
   def importedImplicits(implicit ctx: Context): List[ImplicitRef] = {
     val pre = site
-    if (isWildcardImport) {
-      val refs = pre.implicitMembers(implicitFlag)
-      if (excluded.isEmpty) refs
-      else refs filterNot (ref => excluded contains ref.name.toTermName)
-    } else
-      for {
+    if (isWildcardImport)
+      pre.implicitMembers(implicitFlag).flatMap { ref =>
+        val name = ref.name.toTermName
+        if (excluded.contains(name)) Nil
+        else {
+          val renamed = forwardMapping(ref.name)
+          if (renamed == ref.name) ref :: Nil
+          else if (renamed != null) new RenamedImplicitRef(ref, renamed) :: Nil
+          else if (!impliedBound.exists || (ref <:< impliedBound)) ref :: Nil
+          else Nil
+        }
+      }
+    else
+      for
         renamed <- reverseMapping.keys
         denot <- pre.member(reverseMapping(renamed)).altsWith(_ is implicitFlag)
-      } yield {
+      yield {
         val original = reverseMapping(renamed)
         val ref = TermRef(pre, original, denot)
         if (renamed == original) ref
```

Comment:
Nice `for` syntax 👍

Reply:
Hum ... It doesn't look like valid Scala code to me, actually :-s Maybe with the indentation-based syntax, but that shouldn't be enabled there? Is the parser too lenient?